docs/proposals/11-cluster-health.md

```diff
@@ -0,0 +1,205 @@
+- Gardener
+  - The `probe` [sub-command](https://github.com/gardener/dependency-watchdog/blob/master/cmd/probe.go) of the `dependency-watchdog` probes the external and internal endpoints of `kube-apiserver` of all the shoots (hosted on the seed cluster).
+  If it finds that any of the external endpoints are not healthy while the internal endpoints are, then it [scales down](https://github.com/gardener/gardener/blob/master/charts/seed-bootstrap/charts/dependency-watchdog/templates/probe-configmap.yaml#L22) the corresponding shoot's `kube-controller-manager` to `0`.
+  This is to prevent the `NodeLifecycle` controller of the `kube-controller-manager` from [marking](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L860) all the `nodes` and their `pods` in the cluster as `NotReady` because the cause of the `nodes` being not ready has probably nothing to do with the `nodes` themselves.
+  It is probably because the `kube-apiserver` endpoint is not reachable externally due to loadbalancer or some other network issue.
+  - The `root` [command](https://github.com/gardener/dependency-watchdog/blob/master/cmd/root.go) of the `dependency-watchdog` watches the Kubernetes `endpoints` of `etcd` and `kube-apiserver` of all the shoots (hosted in the seed cluster).
+  If any of the endpoints transition from unhealthy to healthy it restarts (deletes) the corresponding [downstream](https://github.com/gardener/gardener/blob/master/charts/seed-bootstrap/charts/dependency-watchdog/templates/endpoint-configmap.yaml#L41) [dependent](https://github.com/gardener/gardener/blob/master/charts/seed-bootstrap/charts/dependency-watchdog/templates/endpoint-configmap.yaml#L16) `pods` but only if they are in `CrashLoopBackoff`.
+  This is to speed up the cluster recovery from `etcd` or `kube-apiserver` unavailability.
+  - The `etcd` `readinessProbe` [points](https://github.com/gardener/gardener/blob/master/charts/seed-controlplane/charts/etcd/templates/etcd-statefulset.yaml#L43) to the health of the [etcd-backup-restore](https://github.com/gardener/etcd-backup-restore) [sidecar](https://github.com/gardener/gardener-extensions/blob/master/pkg/webhook/controlplane/etcd.go#L44).
+  This makes the `etcd` of the control-plane of the shoot clusters unavailable if the backup infrastructure is unavailable for any reason (even if the `etcd` container itself is healthy).
+  - The `machine-controller-manager` can in principle (it currently doesn't) notice issues in the cloud-provider that other Kubernetes and or Gardener controllers do not notice such as machines not joining in the given timeout period, cloud-provider API requests failing etc.
+
+The distributed nature of such information and action is in itself not a *bad thing*.
```

Comment:
I think I should give more context explicitly in the [motivation](https://github.com/gardener/gardener/pull/1797/files#diff-dc2b4cab766fb3ad06a4e681a236b570R33) section. I will do that.

But to explain briefly here, the conditions that I have in mind (as you can see from the samples I have [given](https://github.com/gardener/gardener/pull/1797/files#diff-dc2b4cab766fb3ad06a4e681a236b570R104) [below](https://github.com/gardener/gardener/pull/1797/files#diff-dc2b4cab766fb3ad06a4e681a236b570R147) are those that need some specific action that can be taken either to rectify the issue or to mitigate further damage to the cluster. Such actions are better taken by small extensions or controllers (which could be aggregated into controller managers), rather than the `gardenlet`. This is because gardenlet can take only generic actions during the reconciliation. Specific actions I referred to may not come under the umbrella of simple reconciliation of the deployed components. Some actions might have to be taken by existing controllers such as the `machine-controller-manager` (e.g. to avoid or slow down the replacement of unhealthy `nodes`), some might involve cloud provider specific actions etc.

Viewed in this context, the shoot care controller maintains the conditions in the `Shoot` resource in the `garden` cluster. This is suitable for external consumption. But only the `gardenlet` is supposed to access the `garden` cluster. I am not sure if it is a good idea if other controllers and extensions in the seed start accessing the `garden` cluster. That would be an anti-pattern.